wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/EnvironmentParser.java

```diff
@@ -78,36 +85,79 @@ public CheServicesEnvironmentImpl parse(Environment environment) throws IllegalA
+                throw new IllegalArgumentException(format("Environment type '%s' is not supported. " +
+                                                          "Supported environment types: %s",
+                                                          envType,
+                                                          Joiner.on(", ").join(types)));
         }
 
-        composeEnvironment.getServices().forEach((name, service) -> {
+        cheServicesEnvironment.getServices().forEach((name, service) -> {
             ExtendedMachine extendedMachine = environment.getMachines().get(name);
-            if (extendedMachine != null &&
-                extendedMachine.getAttributes() != null &&
-                extendedMachine.getAttributes().containsKey("memoryLimitBytes")) {
-
-                try {
-                    service.setMemLimit(Long.parseLong(extendedMachine.getAttributes().get("memoryLimitBytes")));
-                } catch (NumberFormatException e) {
-                    throw new IllegalArgumentException(
-                            format("Value of attribute 'memoryLimitBytes' of machine '%s' is illegal", name));
-                }
+            if (extendedMachine != null) {
+                normalizeMachine(name, service, extendedMachine);
             }
         });
 
-        return composeEnvironment;
+        return cheServicesEnvironment;
+    }
+
+    private void normalizeMachine(String name, CheServiceImpl service, ExtendedMachine extendedMachine) {
+        if (extendedMachine.getAttributes() != null &&
+            extendedMachine.getAttributes().containsKey("memoryLimitBytes")) {
+
+            try {
+                service.setMemLimit(Long.parseLong(extendedMachine.getAttributes().get("memoryLimitBytes")));
+            } catch (NumberFormatException e) {
+                throw new IllegalArgumentException(
+                        format("Value of attribute 'memoryLimitBytes' of machine '%s' is illegal", name));
+            }
+        }
+        if (service.getExpose() != null) {
+            service.setExpose(service.getExpose()
+                                     .stream()
+                                     .map(expose -> expose.contains("/") ?
+                                                    expose :
+                                                    expose + "/tcp")
+                                     .collect(toList()));
+        }
+        if (extendedMachine.getServers() != null) {
+            extendedMachine.getServers().forEach((serverRef, serverConf) -> {
+                String normalizedPort = serverConf.getPort().contains("/") ?
+                                        serverConf.getPort() :
+                                        serverConf.getPort() + "/tcp";
+
+                service.getExpose().add(normalizedPort);
+
+                service.getLabels().put(SERVER_CONF_LABEL_PREFIX +
```

Comment:
labels are part of non-rest code, so it returns empty collections always and proxy doesn't affect it